File: 0778-swim-in-rising-water/0778-swim-in-rising-water.java
class Tuple{
    int time;
    int row;
    int col;

    public Tuple(int time,int row,int col){
        this.time = time;
        this.row = row;
        this.col = col;
    }
}

class Solution {
    public int swimInWater(int[][] grid) {

        int n = grid.length;

        PriorityQueue<Tuple> pq = new PriorityQueue<>((x,y)->Integer.compare(x.time,y.time));

        int vis[][] = new int[n][n];

        pq.add(new Tuple(grid[0][0],0,0));

        while(!pq.isEmpty()){

            Tuple it = pq.poll();
            int currTime = it.time;
            int r = it.row;
            int c = it.col;

            if(r==n-1 && c==n-1) return currTime;

            int dr[] = {-1,0,1,0};
            int dc[] = {0,1,0,-1};

            for(int i=0;i<4;i++){
                int nrow = r + dr[i];
                int ncol = c + dc[i];

                if(nrow>=0 && nrow<n && ncol>=0 && ncol<n && vis[nrow][ncol] ==0){
                    vis[nrow][ncol] = 1;
                    pq.add(new Tuple(Math.max(grid[nrow][ncol],currTime),nrow,ncol));
                }
            }
        }

        return -1;
        
    }
}

File: 0778-swim-in-rising-water/0778-swim-in-rising-water.cpp
class Solution {
public:
    int swimInWater(vector<vector<int>>& grid) {
        int n = grid.size();

        using Node = pair<int,pair<int,int>>;
        priority_queue<
                    Node,vector<Node>,greater<Node> > pq;

        vector<vector<int>> vis(n,vector<int>(n,0));

        pq.push({grid[0][0],{0,0}});

        while(!pq.empty()){
            int currTime = pq.top().first;
            int row = pq.top().second.first;
            int col = pq.top().second.second;

            pq.pop();

            if(row == n-1 && col == n-1) return currTime;

            int dr[] = {-1,0,1,0};
            int dc[] = {0,1,0,-1};

            for(int i=0;i<4;i++){
                int nrow = row + dr[i];
                int ncol = col + dc[i];

                if(nrow>=0 && nrow<n && ncol>=0 && ncol<n && !vis[nrow][ncol]){
                    vis[nrow][ncol] = 1;
                    pq.push({max(grid[nrow][ncol],currTime),{nrow,ncol}});
                }
            }
        }

        return -1;

    }
};